source/common/ssl/context_impl.cc

```diff
@@ -31,37 +32,64 @@ int ContextImpl::sslContextIndex() {
-  RELEASE_ASSERT(rc == 1, "");
-
-  if (!SSL_CTX_set_strict_cipher_list(ctx_.get(), config.cipherSuites().c_str())) {
-    std::vector<absl::string_view> ciphers =
-        StringUtil::splitToken(config.cipherSuites(), ":+-![|]", false);
-    std::vector<std::string> bad_ciphers;
-    for (const auto& cipher : ciphers) {
-      std::string cipher_str(cipher);
-      if (!SSL_CTX_set_strict_cipher_list(ctx_.get(), cipher_str.c_str())) {
-        bad_ciphers.push_back(cipher_str);
+    : scope_(scope), stats_(generateStats(scope)), time_source_(time_source) {
+  const auto tls_certificates = config.tlsCertificates();
+  tls_contexts_.resize(std::max(1UL, tls_certificates.size()));
+
+  for (auto& ctx : tls_contexts_) {
+    ctx.ssl_ctx_.reset(SSL_CTX_new(TLS_method()));
+
+    int rc = SSL_CTX_set_ex_data(ctx.ssl_ctx_.get(), sslContextIndex(), this);
+    RELEASE_ASSERT(rc == 1, "");
+
+    rc = SSL_CTX_set_min_proto_version(ctx.ssl_ctx_.get(), config.minProtocolVersion());
+    RELEASE_ASSERT(rc == 1, "");
+
+    rc = SSL_CTX_set_max_proto_version(ctx.ssl_ctx_.get(), config.maxProtocolVersion());
+    RELEASE_ASSERT(rc == 1, "");
+
+    if (!SSL_CTX_set_strict_cipher_list(ctx.ssl_ctx_.get(), config.cipherSuites().c_str())) {
+      std::vector<absl::string_view> ciphers =
+          StringUtil::splitToken(config.cipherSuites(), ":+-![|]", false);
+      std::vector<std::string> bad_ciphers;
+      for (const auto& cipher : ciphers) {
+        std::string cipher_str(cipher);
+        if (!SSL_CTX_set_strict_cipher_list(ctx.ssl_ctx_.get(), cipher_str.c_str())) {
+          bad_ciphers.push_back(cipher_str);
+        }
       }
+      throw EnvoyException(fmt::format("Failed to initialize cipher suites {}. The following "
+                                       "ciphers were rejected when tried individually: {}",
+                                       config.cipherSuites(), StringUtil::join(bad_ciphers, ", ")));
     }
-    throw EnvoyException(fmt::format("Failed to initialize cipher suites {}. The following "
-                                     "ciphers were rejected when tried individually: {}",
-                                     config.cipherSuites(), StringUtil::join(bad_ciphers, ", ")));
-  }
 
-  if (!SSL_CTX_set1_curves_list(ctx_.get(), config.ecdhCurves().c_str())) {
-    throw EnvoyException(fmt::format("Failed to initialize ECDH curves {}", config.ecdhCurves()));
+    if (!SSL_CTX_set1_curves_list(ctx.ssl_ctx_.get(), config.ecdhCurves().c_str())) {
+      throw EnvoyException(fmt::format("Failed to initialize ECDH curves {}", config.ecdhCurves()));
+    }
+
+    // TODO(htuch): replace with SSL_IDENTITY when we have this as a means to do multi-cert in
+    // BoringSSL.
+    SSL_CTX_set_select_certificate_cb(
+        ctx.ssl_ctx_.get(), [](const SSL_CLIENT_HELLO* client_hello) -> ssl_select_cert_result_t {
+          ContextImpl* context_impl =
+              static_cast<ContextImpl*>(SSL_get_app_data(client_hello->ssl));
```

Comment:
Note: it's fine if you prefer using `SSL_CTX_get_app_data()` instead of `SSL_CTX_get_ex_data(sslContextIndex)`, since the former is shorter-form and we only store one item anyway, but if that's the case, then please migrate existing use of `SSL_CTX_get_ex_data(sslContextIndex)` to `SSL_CTX_get_app_data()` and remove `sslContextIndex()` (...or I can do that, if you prefer), but let's keep it on `SSL_CTX` and not `SSL`.